### DISCUSSION TITLE
Building ISG pipeline to collect all types if type has changed or changed its constructor

### DISCUSSION THREAD

**Discussion by lsoft** [dotnet/roslyn#71648] [Q&A] [answered]
Hello.

Imagine, I need an ISG which collects all constructors inside every type of the assembly and builds somethings like this:
```C#
public static class ConstructorInfoProvider
{
   public string GetAllConstructorsInfo() => @"
global::MyType(global::System.String a, global::System.Int32 b);
...
"; //this is constructor list
}
```

I understand that last step of this ISG will be expensive since ISG produced a big portion of the code. I need to regenerate this code 

1. in case of any type added\removed to the assembly
2. changed type's name or changed type's namespace
3. changed a declaration part of any of its constructors (not the body of constructor), including `base(...)` and `this(...)`.

Usually, programmers do not change the constructor or added types often, so, I hope that last (expensive) step of ISG pipeline will be called rarely.

My question is how to organize ISG pipeline efficiently?

Looks like I can build the following pipeline:
```C#
            var step1 = context.CompilationProvider.SelectMany(
                (compilation, ct) =>
                {
                    var result = new List<INamedTypeSymbol>(
                        compilation.Assembly.TypeNames.Count //preallocate
                        );

                    result.AddRange(compilation.Assembly.GlobalNamespace.GetAllTypes());

                    return result.ConverAll(r => new EquatableConstructorsInfo(r)).ToImmutableArray();
                }
                );
               //then Collect info from step1
               //then context.RegisterSourceOutput collected info
```

where `GetAllTypes` is a helper to get all types inside an assembly, and `EquatableConstructorsInfo` is a class/struct (which can be compared to itself to enable ISG pipeline cache) about type constructors (and does not contains a `Compilation` or any `ISymbol`). So, if any of type's constructor has changed, `EquatableConstructorsInfo` will be different, and `Collect` will be called to rebuild the ISG result.

but I'm afraid that this first step is very expensive, since ISG need to iterate all assembly's types. Are these my considerations true? Or this is not expensive as much?

The other approach: is it possible to implement the same pipeline starting from `SyntaxProvider`? I can't figure out it due to small experience in ISG area... will I have any profit with `SyntaxProvider` in comparison with `CompilationProvider`?

Thanks!

**Comment by CyrusNajmabadi** [comment, marked as answer]
> Imagine, I need an ISG which collects all constructors inside every type

This would be extremely expensive in any world.  Consider moving to a model where the relevant constructors are marked with an attribute. WTH that, you can use ForAttributeWithMetadataName to easily get the results indexed and provided to you with minimal cost and effort
